File: Timing.cpp
#include "Timing.hpp"
#include <iostream>
#include <stdexcept>



// Конструктор копирования
Timing::Timing(const Timing& other) {
    for (const auto& trip : other.tripList) {
        // новые копии поездок
        tripList.push_back(std::make_shared<Trip>(*trip));
    }
}

void Timing::AddTrip(std::shared_ptr<Trip> trip) {
    try {
        if (!trip) {
            throw std::invalid_argument("Нельзя добавить пустую поездку!");
        }

        // Проверяем дублирование поездок
        for (const auto& existingTrip : tripList) {
            if (existingTrip->GetRoute() == trip->GetRoute() &&
                existingTrip->GetStatus() != "Завершена") {
                throw std::runtime_error("Поездка с маршрутом '" + trip->GetRoute() + "' уже существует!");
            }
        }

        tripList.push_back(trip);
        std::cout << "Поездка добавлена в расписание: " << trip->GetRoute() << "\n";
    }
    catch (const std::exception& e) {
        std::cerr << "Ошибка добавления поездки: " << e.what() << "\n";
        throw;
    }
}

bool Timing::RemoveTrip(const std::string& route) {
    try {
        if (route.empty()) {
            throw std::invalid_argument("Маршрут не может быть пустым!");
        }

        auto it = std::find_if(tripList.begin(), tripList.end(),
            [&route](const std::shared_ptr<Trip>& trip) {
                return trip && trip->GetRoute() == route;
            });

        if (it == tripList.end()) {
            std::cout << "Поездка с маршрутом '" << route << "' не найдена!\n";
            return false;
        }

        // Освобождаем ресурсы
        auto bus = (*it)->GetBus();
        if (bus) {
            bus->ChangeAvailBus(true);
            std::cout << "Автобус " << bus->GetCode() << " освобожден\n";
        }

        auto driver = (*it)->GetDriver();
        if (driver) {
            driver->change_work_avail(true);
            std::cout << "Водитель " << driver->GetFullName() << " освобожден\n";
        }

        // Удаляем поездку
        tripList.erase(it);
        std::cout << "Поездка '" << route << "' успешно удалена!\n";
        return true;
    }
    catch (const std::exception& e) {
        std::cerr << "Ошибка удаления поездки: " << e.what() << "\n";
        return false;
    }
}

void Timing::DisplayAllTrips() const {
    std::cout << "=== Все поездки (" << tripList.size() << ") ===\n";
    if (tripList.empty()) {
        std::cout << "Расписание пусто\n";
    }
    else {
        for (const auto& trip : tripList) {
            std::cout << "• " << trip->GetRoute()
                << " - " << trip->GetStatus()
                << " (" << trip->GetPrice() << " руб.)\n";
        }
    }
    std::cout << "==========================\n";
}

void Timing::DisplayActiveTrips() const {
    std::cout << "=== Активные поездки ===\n";
    int count = 0;
    for (const auto& trip : tripList) {
        if (trip->GetStatus() == "Запланирована" || trip->GetStatus() == "В пути") {
            std::cout << "- " << trip->GetRoute()
                << " - " << trip->GetStatus() << "\n";
            count++;
        }
    }
    if (count == 0) {
        std::cout << "Нет активных поездок\n";
    }
    std::cout << "==========================\n";
}

void Timing::DisplayCompletedTrips() const {
    std::cout << "=== Завершенные поездки ===\n";
    int count = 0;
    for (const auto& trip : tripList) {
        if (trip->GetStatus() == "Завершена") {
            std::cout << "- " << trip->GetRoute()
                << " - " << trip->GetPrice() << " руб.\n";
            count++;
        }
    }
    if (count == 0) {
        std::cout << "Нет завершенных поездок\n";
    }
    std::cout << "==========================\n";
}

void Timing::DisplayPlannedTrips() const {
    std::cout << "=== Запланированные поездки ===\n";
    int count = 0;
    for (const auto& trip : tripList) {
        if (trip->GetStatus() == "Запланирована") {
            std::cout << "• " << trip->GetRoute()
                << " - " << trip->GetPrice() << " руб.\n";
            count++;
        }
    }
    if (count == 0) {
        std::cout << "Нет запланированных поездок\n";
    }
    std::cout << "==========================\n";
}

std::shared_ptr<Trip> Timing::ChoseTrip() {
    try {
        DisplayActiveTrips();

        if (tripList.empty()) {
            throw std::runtime_error("Нет доступных поездок!");
        }

        std::string selectedRoute;
        std::cout << "Введите маршрут поездки: ";
        std::cin.ignore();
        std::getline(std::cin, selectedRoute);

        if (selectedRoute.empty()) {
            throw std::invalid_argument("Маршрут не может быть пустым!");
        }

        for (auto& trip : tripList) {
            if (trip->GetRoute() == selectedRoute &&
                (trip->GetStatus() == "Запланирована" || trip->GetStatus() == "В пути")) {
                std::cout << "Выбрана поездка: " << selectedRoute << "\n";
                return trip;  // Возвращаем умный указатель
            }
        }

        throw std::runtime_error("Активная поездка с маршрутом '" + selectedRoute + "' не найдена!");
    }
    catch (const std::exception& e) {
        std::cerr << "Ошибка выбора поездки: " << e.what() << "\n";
        return nullptr;
    }
}

File: Timing.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Timing {
    private List<Trip> tripList;
    

    public Timing() {
        this.tripList = new ArrayList<>();
    }

    public void addTrip(Trip trip) {
        try {
            if (trip == null) {
                throw new IllegalArgumentException("Trip cannot be null!");
            }
            
            this.tripList.add(trip);
            System.out.println("Trip added to schedule: " + trip.getRoute());
            
        } catch (IllegalArgumentException e) {
            System.err.println("Error adding trip: " + e.getMessage());
            throw e;
        }
    }

    public void displayAllTrips() {
        System.out.println("=== All trips (" + this.tripList.size() + ") ===");
        if (this.tripList.isEmpty()) {
            System.out.println("Schedule is empty");
        } else {
            for (Trip trip : this.tripList) {
                System.out.println("• " + trip.getRoute()
                    + " - " + trip.getStatus()
                    + " (" + trip.getPrice() + " rub.)");
            }
        }
        System.out.println("==========================");
    }

    public void displayActiveTrips() {
        System.out.println("=== Active trips ===");
        int count = 0;
        for (Trip trip : this.tripList) {
            if (trip.getStatus().equals("Planned") || trip.getStatus().equals("In progress")) {
                System.out.println("- " + trip.getRoute()
                    + " - " + trip.getStatus());
                count++;
            }
        }
        if (count == 0) {
            System.out.println("No active trips");
        }
        System.out.println("==========================");
    }

    public void displayCompletedTrips() {
        System.out.println("=== Completed trips ===");
        int count = 0;
        for (Trip trip : this.tripList) {
            if (trip.getStatus().equals("Completed")) {
                System.out.println("- " + trip.getRoute()
                    + " - " + trip.getPrice() + " rub.");
                count++;
            }
        }
        if (count == 0) {
            System.out.println("No completed trips");
        }
        System.out.println("==========================");
    }

    public void displayPlannedTrips() {
        System.out.println("=== Planned trips ===");
        int count = 0;
        for (Trip trip : this.tripList) {
            if (trip.getStatus().equals("Planned")) {
                System.out.println("• " + trip.getRoute()
                    + " - " + trip.getPrice() + " rub.");
                count++;
            }
        }
        if (count == 0) {
            System.out.println("No planned trips");
        }
        System.out.println("==========================");
    }

    public Trip chooseTrip() {
        try {
            this.displayActiveTrips();

            if (this.tripList.isEmpty()) {
                throw new IllegalStateException("No available trips!");
            }

            Scanner scanner = new Scanner(System.in);
            System.out.print("Enter trip route: ");
            String selectedRoute = scanner.nextLine();
            
            if (selectedRoute == null || selectedRoute.trim().isEmpty()) {
                throw new IllegalArgumentException("Route cannot be empty!");
            }

            for (Trip trip : this.tripList) {
                if (trip.getRoute().equals(selectedRoute) &&
                    (trip.getStatus().equals("Planned") || trip.getStatus().equals("In progress"))) {
                    System.out.println(" Selected trip: " + selectedRoute);
                    return trip;
                }
            }

            throw new IllegalStateException("Active trip with route '" + selectedRoute + "' not found!");
            
        } catch (IllegalArgumentException | IllegalStateException e) {
            System.err.println("Error choosing trip: " + e.getMessage());
            return null;
        }
    }

    
    public boolean removeTrip(String route) {
        try {
            if (route == null || route.trim().isEmpty()) {
                throw new IllegalArgumentException("Route cannot be null or empty!");
            }
            
            // Используем Iterator для безопасного удаления
            java.util.Iterator<Trip> iterator = this.tripList.iterator();
            while (iterator.hasNext()) {
                Trip trip = iterator.next();
                if (trip.getRoute().equals(route)) {
                    // Проверяем, можно ли удалить поездку (только запланированные можно удалять)
                    if ("Planned".equals(trip.getStatus())) {
                        iterator.remove();
                        System.out.println("Trip '" + route + "' removed from schedule.");
                        return true;
                    } else {
                        throw new IllegalStateException("Cannot delete trip '" + route + "' because it's " + trip.getStatus() + ". Only planned trips can be deleted.");
                    }
                }
            }
            
            System.out.println("Trip with route '" + route + "' not found.");
            return false;
            
        } catch (IllegalArgumentException | IllegalStateException e) {
            System.err.println("Error removing trip: " + e.getMessage());
            throw e;
        }
    }

    // Геттер
    public List<Trip> getTripList() {
        return this.tripList;
    }
}